ddtrace/internal/datadog/profiling/dd_wrapper/src/uploader.cpp

```diff
@@ -0,0 +1,116 @@
+#include "uploader.hpp"
+#include "libdatadog_helpers.hpp"
+
+using namespace Datadog;
+
+void
+DdogProfExporterDeleter::operator()(ddog_prof_Exporter* ptr) const
+{
+    ddog_prof_Exporter_drop(ptr);
+}
+
+Uploader::Uploader(std::string_view _url, ddog_prof_Exporter* _ddog_exporter)
+  : ddog_exporter{ _ddog_exporter }
+  , url{ _url }
+{
+}
+
+bool
+Uploader::upload(ddog_prof_Profile& profile)
+{
+    // Serialize the profile
+    ddog_prof_Profile_SerializeResult result = ddog_prof_Profile_serialize(&profile, nullptr, nullptr, nullptr);
+    if (result.tag != DDOG_PROF_PROFILE_SERIALIZE_RESULT_OK) {
+        errmsg = err_to_msg(&result.err, "Error serializing pprof");
+        ddog_Error_drop(&result.err);
+        std::cerr << errmsg << std::endl;
+        return false;
+    }
+    ddog_prof_EncodedProfile* encoded = &result.ok;
+
+    // If we have any custom tags, set them now
+    ddog_Vec_Tag tags = ddog_Vec_Tag_new();
+    add_tag(tags, ExportTagKey::runtime_id, runtime_id, errmsg);
```

Comment:
What if this fails?

Reply:
Factored to now add a more descriptive error, indicating which values failed and why.